Endert Guerrero 

Módulo: 1

Asignatura: Introducción al Machine Learning

Docente Online: Freddy Hernan Sanchez Vallejos

# Introducción



# Librerias

Importare las librerías necesarias para realizar el modulo.

In [92]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Importación del dataset 

Importare el Dataset haciendo uso de pandas 

In [93]:
df = pd.read_excel("c:/Users/aleja/OneDrive/Escritorio/titanic_M1.xlsx")
df.head(11)

pclass  survived                                             name     sex  \
0      1.0         1                    Allen, Miss. Elisabeth Walton  female   
1      1.0         1                   Allison, Master. Hudson Trevor    male   
2      1.0         0                     Allison, Miss. Helen Loraine  female   
3      1.0         0             Allison, Mr. Hudson Joshua Creighton    male   
4      1.0         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5      1.0         1                              Anderson, Mr. Harry    male   
6      1.0         1                Andrews, Miss. Kornelia Theodosia  female   
7      1.0         0                           Andrews, Mr. Thomas Jr    male   
8      1.0         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9      1.0         0                          Artagaveytia, Mr. Ramon    male   
10     1.0         0                           Astor, Col. John Jacob    male   

        age  sibsp  parch    ticket       fare    cabin embarked boat   body  \
0   29.0000      0      0     24160  2113375.0       B5        S    2    NaN   
1    0.9167      1      2    113781  1515500.0  C22 C26        S   11    NaN   
2    2.0000      1      2    113781  1515500.0  C22 C26        S  NaN    NaN   
3   30.0000      1      2    113781  1515500.0  C22 C26        S  NaN  135.0   
4   25.0000      1      2    113781  1515500.0  C22 C26        S  NaN    NaN   
5   48.0000      0      0     19952   265500.0      E12        S    3    NaN   
6   63.0000      1      0     13502   779583.0       D7        S   10    NaN   
7   39.0000      0      0    112050        0.0      A36        S  NaN    NaN   
8   53.0000      2      0     11769   514792.0     C101        S    D    NaN   
9   71.0000      0      0  PC 17609   495042.0      NaN        C  NaN   22.0   
10  47.0000      1      0  PC 17757  2275250.0  C62 C64        C  NaN  124.0   

                          home.dest  
0                      St Louis, MO  
1   Montreal, PQ / Chesterville, ON  
2   Montreal, PQ / Chesterville, ON  
3   Montreal, PQ / Chesterville, ON  
4   Montreal, PQ / Chesterville, ON  
5                      New York, NY  
6                        Hudson, NY  
7                       Belfast, NI  
8               Bayside, Queens, NY  
9               Montevideo, Uruguay  
10                     New York, NY

In [94]:
shape1 = df.shape
print(shape1)

(1219, 14)


# Ítem 1

Los datos faltantes presentes en los Dataset son variados, hare uso de dos métodos, el primero que es ".dtypes" me describirá el tipo de datos de dicha columna, por otro lado el método ".isnull" en conjunto con ".sum" me ayudan a visualizar el numero de datos nulos de cada columna.

también podemos usar el método ".isna" y realizar la misma acción del método ".isnull", para el caso usare ".isnull" ya que nos arrojara el mismo resultado.

In [95]:
datos_nulos = df.isnull().sum()
df.dtypes

pclass       float64
survived       int64
name          object
sex           object
age          float64
sibsp          int64
parch          int64
ticket        object
fare         float64
cabin         object
embarked      object
boat          object
body         float64
home.dest     object
dtype: object

La descripción del tipo de dato nos puede ayudar a entender mejor los datos, podemos ver que algunos son floats y otros son integers, asi como también hay strings.

Gracias a esto podemos llegar a conclusiones concretas de a que corresponde cada dato, por ejemplo en la rubrica no se nos habla de a que corresponde el dato "body", el cual es una columna bastante particular a primera vista porque tiene bastantes datos NaN (Not a Number) y en los casos en que hay un dato presente que a pesar de ser de tipo float podemos ver que son enteros, notamos que existe una correlación con sí el pasajero sobrevivió, solo hay datos presentes en los casos en los que esto no es así, entonces podemos llegar a la conclusion de que este valor corresponde al numero de asignado al cadaver, por eso vemos tantos NaN, ya que muchos cuerpos no fueron recuperados o sobrevivieron.

In [96]:
print(datos_nulos)

pclass         76
survived        0
name            0
sex             0
age           239
sibsp           0
parch           0
ticket          0
fare            0
cabin         927
embarked        2
boat          751
body         1110
home.dest     475
dtype: int64


Al emplear el método ".isnull", se revela que en la columna "pclass", la cual indica la clase en la que el pasajero embarcó, hay 76 datos nulos. Asimismo, se identifican 239 datos faltantes en la columna "age", que representa la edad del pasajero. Se presume que esta ausencia de datos podría deberse, en su mayoría, a personas que no lograron sobrevivir.

La columna "cabin", que describe el tipo de cabina del pasajero, presenta 927 datos nulos. Además, se observan dos datos faltantes en la columna "embarked", la cual indica la ciudad donde los pasajeros embarcaron. Respecto a la columna "boat", que indica el número de bote en el que lograron desembarcar durante el accidente, encontramos 751 datos nulos.

La columna "body" registra 1110 datos nulos, pero como se explicó anteriormente, esta información corresponde al número de cadáver después de ser recuperado. Es normal que haya una gran cantidad de datos nulos, ya que muchos corresponden a cadáveres no recuperados o personas que lograron sobrevivir.

Finalmente, la columna 'home.dest' refleja el lugar de origen del pasajero.

# Ítem 2 

Los valores nulos en la columna 'pclass' fueron rellenados asignándoles la clase 2. Esto se realizó utilizando el método fillna() en la columna 'pclass' del DataFrame, con el valor 2 especificado como reemplazo.

In [97]:
df["pclass"].fillna(2, inplace = True)
df.isna().sum()

pclass          0
survived        0
name            0
sex             0
age           239
sibsp           0
parch           0
ticket          0
fare            0
cabin         927
embarked        2
boat          751
body         1110
home.dest     475
dtype: int64

Finalmente para asegurarse de que los cambios se aplicaron correctamente, se volvió a verificar la presencia de valores nulos después de la imputación usando ".isna".

# Ítem 3

Para comprender la diversidad de valores en la columna 'Sex', se imprimieron los valores únicos utilizando el método unique().

In [98]:
print(df["sex"].unique())

['female' 'male' 'Hombre' 'Mujer']


Esto proporcionó una visión inicial de los valores presentes en la columna 'Sex'.

Luego se definieron los valores deseados para 'Sex', que son "male" y "female", y se almacenaron en una lista llamada "sex".

Se filtraron los datos de la columna 'Sex' para incluir solo los valores definidos en la lista "sex".

In [99]:
sex = ["female","male"]

df = df[df["sex"].isin(sex)]

print(df["sex"].unique())

['female' 'male']


Esto resultó en un DataFrame modificado que contiene únicamente las filas donde 'Sex' es "male" o "female".

Para evaluar el impacto en el tamaño del DataFrame después de la imputación, se imprimió el tamaño del DataFrame, esto proporcionó información sobre la cantidad de filas y columnas en el DataFrame después de realizar los cambios en la columna 'Sex' que en comparación al inicio nos da que hemos eliminado una cierta cantidad de filas.

In [108]:
shape2 = df.shape

diferencia_filas = shape1[0] - shape2[0]

print(f"Diferencia de filas con respecto al DataFrame inicial después de imputar los datos: {diferencia_filas}")

Diferencia de filas con respecto al DataFrame inicial después de imputar los datos: 437


# Ítem 4

Se imprimieron las estadísticas descriptivas de la columna 'age' para entender la distribución inicial de los datos, incluyendo el promedio.

Esto proporcionó información sobre la media y el numero de datos no nulos que existen hasta el momento.

In [101]:
print(df["age"].describe())

count    627.000000
mean      30.448432
std       14.426633
min        0.333300
25%       21.000000
50%       28.000000
75%       39.000000
max       80.000000
Name: age, dtype: float64


Se calculó el promedio de edad utilizando el método mean() en la columna 'age' y se almacenó en la variable age_mean.

Este paso permitió obtener el promedio de edad de aquellos que tenían el dato disponible.

In [102]:
age_mean = df["age"].mean()
df["age"].fillna(age_mean, inplace=True)

La imputación con el promedio garantizó que los valores faltantes en 'age' fueran sustituidos por la media de los valores conocidos.

Finalmente se volvieron a imprimir las estadísticas descriptivas de la columna 'age' para verificar cómo afectó la imputación al conjunto de datos, asi podemos confirmar que el numero de datos no nulos creció, obteniendo asi 155 datos que fueron reemplazados por la media de edad que es 30.44 años.

In [103]:
print(df["age"].describe())



count    782.000000
mean      30.448432
std       12.915958
min        0.333300
25%       23.000000
50%       30.448432
75%       35.000000
max       80.000000
Name: age, dtype: float64


# Ítem 5


Se imprimen las estadísticas descriptivas de la columna 'embarked' para verificar mas adelante como afectara la asignación de "S" al conjunto de datos.

In [104]:
print(df["embarked"].describe())

count     781
unique      3
top         S
freq      561
Name: embarked, dtype: object


Se asignó el valor "S" a aquellos datos en la columna 'embarked' que se encontraban con valores vacíos. Esto se hizo mediante el método "fillna()".

Luego se volvieron a imprimir las estadísticas descriptivas de la columna 'embarked' para verificar cómo afectó la asignación de "S" al conjunto de datos.

In [105]:
embarked_city = "S"
df["embarked"].fillna(embarked_city, inplace=True)

print(df["embarked"].describe())

count     782
unique      3
top         S
freq      562
Name: embarked, dtype: object


Este paso confirmó que los valores vacíos fueron reemplazados por "S" y ahora tenemos 782 no nulos.

# Conclusión

La preparación de datos desempeña un papel crucial en la construcción de modelos de machine learning. Garantizar la coherencia, integridad y calidad de los datos es esencial para obtener resultados precisos y confiables. En este proceso por ejemplo abordamos valores faltantes en las columnas 'age' y 'embarked', asegurando que nuestro conjunto de datos esté listo para ser utilizado en un modelo. La atención a la preparación de datos es fundamental para maximizar la eficacia y la interpretación de los algoritmos de machine learning.

Para finalizar mostraremos en pantalla que ya no existen datos nulos en las columnas en las que queremos que no existan estos.

In [106]:
print(df.isnull().sum())

pclass         0
survived       0
name           0
sex            0
age            0
sibsp          0
parch          0
ticket         0
fare           0
cabin        617
embarked       0
boat         560
body         698
home.dest    325
dtype: int64


In [109]:
df.shape

(782, 14)